### DISCUSSION TITLE
How to generate plots showing % PC contribution of markers (PCA results)?

### DISCUSSION THREAD

**Discussion by denvercal1234GitHub** [ImmuneDynamics/Spectre#97] [Help] [answered]
Hi there, 

The guidance here (https://wiki.centenary.org.au/display/SPECTRE/Tutorial%3A+which+markers+to+use+for+clustering) explains very well the impact of selecting certain markers for clustering by dropping out/in certain markers. 

Nevertheless, related to Discussion #30 (question 3), I wonder if we could generate a plot to visualize which protein markers contribute the most to the PCA results? I hope to have a ranked list of the markers, then use biological knowledge for further selection. 

Thank you for your help!

**Comment by tomashhurst** [comment]
@denvercal1234GitHub indeed we can! @Felixillion do you have a tutorial or protocol for your PCA run handy?

**Reply by Felixillion** [reply]
@denvercal1234GitHub @tomashhurst I'm working on a `run.pca` workflow now; will try and have it ready next week. You can do it now with `scree.plot = TRUE` and `variable.contribution = TRUE`, but the workflow should make things clearer. I've added a lot of options that make sense to me (but not necessarily anyone else)...

**Reply by denvercal1234GitHub** [reply]
Thank you so much @Felixillion. If possible and you have time (I know 😅), would you mind letting us know here when the workflow is ready?

**Reply by Felixillion** [reply]
@denvercal1234GitHub As promised, here it is: https://immunedynamics.io/spectre/tutorials/pca.
Let me know if you have any other questions!

**Reply by denvercal1234GitHub** [reply]
Hi @Felixillion - Thank you for a very engaging and useful tutorial on running PCA within `Spectre`. 

I am a bit confused of what being plotted here. It appeared in the workflow example, for each `FileName`, it calculated the sum of expression values across all markers for a given cluster, then divided it by the sum in expression values across all markers across all clusters within a `FileName`? This calculation resulted in 7 mc groups per `FileName`. These mc groups were then plotted onto a PCA space to represent how expression level of all markers from each cluster relates to that of other clusters?

**Would you mind helping me though understand how I would adopt this PCA workflow for the 2 use cases below? Specifically, I am not sure how to compute my "mc" grouping in these scenarios.**

(1) To identify which markers contribute most variation to the clustering? The Aim is to select which markers I should use or could exclude in the clustering
(2) To identify which FCS files might group together away from the rest of the FCS files? The Aim is to identify potential troubling FCS files (i.e., outlier FCS files). Related to Issue #148

Thanks again for your great work and help.

**Reply by Felixillion** [reply]
Hi @denvercal1234GitHub ! Sorry for my slow reply.

Glad to hear you enjoyed the tutorial!

Regarding your questions:
1. In the tutorial, I've created the PCA based on proportions of each of the clusters. You could run PCA just on marker expression, which will then show you the markers that have high variance across individual cells. In theory this will help identify markers that could separate out cells you could use for clustering, but it may miss markers important for rare populations. Worth trying though to see what results you want.
2. I have done similar things before by first clustering on cells using all markers, then calculating the proportions for each cluster and patient, and then running PCA. In my experiment I had internal controls from the same patient, so on the PCA plot I could see individual samples that were outliers. I could then look at the cluster contribution to get an idea of which clusters were different to then find individual markers that differed between experiments. I mainly used this to quickly see if I should do batch correction (like CytoNorm), and then I could compare before and after alignment to see if there was any difference. Outliers are always hard to identify (especially human data). If you're expecting all samples to be similar, the PCA won't help identify outliers – it's looking for total variance so even if the samples are as close to identical, it will split the groups. If you see one sample further away from the rest, it may be an outlier, but again it's always hard to tell. One option is to cluster on everything but the outliers, do your comparisons, and then use everything for manual gating where you can better control for variability. In this way you're not losing any data. Alternatively you can run batch correction to normalise the data.

I hope that helps!

**Reply by denvercal1234GitHub** [reply]
Hi @Felixillion -- Thank you for your response! Very useful! 

For question 2, when I manually examined all the FCS files (each FCS file had a different antibody for PE), I noticed some FCS files have different staining patterns than the rest of the FCS files, likely because the antibody amount for PE in these files are too much for the marker, and as a result, it influences other channels hence producing the "strange" staining pattern (see below). 

**So I adopted your PCA tutorial to use the distribution of cells per cluster (with PE channel column removed) as a measure to identify the potential outlier FCS files on the PCA plots.** (I expect to see that would group together away from the "expected" files)

Q3. I was able to obtain these PCA plots as below, and it does appear that the files I manually identified as "strange" (Groups = T) separate away from the other group with Dim1. **Do you know how I might compute some quantitative metrics to select the extreme of these "yellow" files (e.g., that yellow one at the top right corner) instead of just by visualizing them on the plot?** The `adonis` only show whether the grouping (strange vs normal in my case) is significant.

Q4. Is there a way to save the plots as png instead of pdf? I want to change the size of the plots as some labels got cropped off.

<img width="692" alt="image" src="https://user-images.githubusercontent.com/62179082/231108710-0e369a77-5a53-4292-931b-9d1a85115093.png">

Thank you again very much for your help!

Related: https://github.com/ImmuneDynamics/Spectre/issues/148

**Reply by Felixillion** [reply, marked as answer]
Hi @denvercal1234GitHub !

The coordinates of each sample on the PCA plot should also have been exported, so you could do some statistical tests to identify outliers based on their location in the plot. (I don't have any recommendations for this as I haven't done it before.)

At the moment files are only saved as PDF, but if you open them in Illustrator any text that's cut off will be present so you should be able to make changes.

**Reply by denvercal1234GitHub** [reply]
Thank you @Felixillion! Do you by chance know a package to see how strong 2 parameters (PE channel vs APC channel data) correlated for flow data?

**Reply by Felixillion** [reply]
@denvercal1234GitHub I'm not aware of a specific package that can do that, but you can look up spread matrices that will give you an idea of how much particular fluorochromes spill into other channels (see figure 2 https://pubmed.ncbi.nlm.nih.gov/32150355/ ). Note the values will be dependent on the flow cytometer being used

**Reply by denvercal1234GitHub** [reply]
Thanks @Felixillion. I aimed to see how well 2 markers agree with each other (i.e., how titled is the diagnonal line in the x by y plot, instead of seeing the spread of one into the other). If Spectre cell.dat has the data for each marker as rowxcolumn with each row being the file, does that mean I can just use standard correlation analysis, e.g., http://www.sthda.com/english/wiki/correlation-test-between-two-variables-in-r? Is it appropriate for flow data?


Also, in your previous response, you mentioned "I could then look at the cluster contribution to get an idea of which clusters were different to then find individual markers that differed between experiments." Did you meant you first see which `mc_` (i.e., which cluster) has the highest absolute value in perhaps `Dim.1` (e.g., mc2 and mc3 in your tutorial). Then, from here, how did you then find the markers that differed between the files? Did you need to first perform differential analysis to get differentially-expressed markers defining each cluster?

<img width="927" alt="Screenshot 2023-04-19 at 12 55 14" src="https://user-images.githubusercontent.com/62179082/233067154-eb713166-e629-4ae7-b6d3-4e9592ee6fdb.png">

**Reply by Felixillion** [reply]
@denvercal1234GitHub when looking at two markers at a single-cell level, it's generally rare to have a perfect correlation (exceptions depend on the markers, like CD3 vs TCR). If there is a strong correlation it's commonly just because of spillover. Regardless, I would think a standard correlation analysis should work. I'd then have a look at the markers and fluorochromes to make sure the results are real and not just because of spillover.

For the next question, yes you can look at the metacluster(s) with the highest values for the dimension(s). Once you have this you can use something like a heatmap to phenotype the metaclusters to give you an idea of what cell subset they are and whether one of their unique markers differs between batches. It's a roundabout way of doing it (as you first cluster then define its phenotype), but this way you can also validate the clusters by making sure they make sense phenotypically.
